File: Backend/src/main/java/com/github/francisfire/anavis/models/AuthCredentials.java
package com.github.francisfire.anavis.models;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@SuppressWarnings("serial")
@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@AllArgsConstructor
@Document(collection = "authcredentials")
public class AuthCredentials implements UserDetails {

	@EqualsAndHashCode.Include
	@Id
	private String mail;
	private String password;
	private Set<Role> roles;

	public enum Role {
		DONOR, OFFICE, ADMIN;
	}

	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		return new ArrayList<GrantedAuthority>(
				roles.stream().map(role -> new SimpleGrantedAuthority(role.name())).collect(Collectors.toList()));
	}

	@Override
	public String getUsername() {
		return mail;
	}

	@Override
	public boolean isAccountNonExpired() {
		return true;
	}

	@Override
	public boolean isAccountNonLocked() {
		return true;
	}

	@Override
	public boolean isCredentialsNonExpired() {
		return true;
	}

	@Override
	public boolean isEnabled() {
		return true;
	}

}

File: Frontend/lib/models/authcredentials.dart
class AuthCredentials {
  String _mail;
  String _password;
  Role _role;
  AuthCredentials(String mail) {
    this._mail = mail;
  }
  AuthCredentials.complete(String mail, String password, Role role) {
    this._mail = mail;
    this._role = role;
    this._password = password;
  }

  String getMail() {
    return _mail;
  }

  String getPassword() {
    return _password;
  }

  void setPassword(String pass) {
    this._password = pass;
  }

  void setRole(Role role) {
    this._role = role;
  }

  Role getRole() {
    return _role;
  }
}

enum Role {
  DONOR,
  OFFICE,
  ADMIN,
}
